Updated macOS paths to OS-canonical paths

Kreeblah opened the pull request:
The default path of `~/.fheroes2` doesn't really match the typical structure on macOS.  This changes the path searched to the following:

`~/Library/Application Support/fheroes2` for data files (the extraction and demo download scripts have also been updated to use this path when extracting/installing data)

`~/Library/Preferences/fheroes2` for configuration/application-generated data

oleg-derevenetz commented:
Hi @Kreeblah @ihhub I have a mixed feelings regarding this PR. First of all, `fheroes2` is not a macOS app - for example, it is not installed in `/Applications` or `~/Applications` and it doesn't have typical macOS app layout. Currently it's more like *nix app (installed via MacPorts or Homebrew), so I'd say that it's better to follow the *nix directory layout guidelines (like `git` for example that stores its settings in `~/.git`), not the macOS guidelines. Second, even if we choose to follow the macOS gudelines, just putting everything in `~/Library` "as is" is not enough, we should also follow naming guidelines (such as `com.horns.and.hoovers.blablabla`), store app preferences in the `~/Library/Preferences/` in plist format (instead of simple text config file `fheroes2.cfg`), etc, etc. I'd say this change is a half solution at best. I'd prefer that `fheroes2` to follow *nix guidelines.

Kreeblah commented:
Fair enough.  I was thinking about doing some work to add an app bundle as a build target next, so would it make sense to close this, and then create a new pull request to cover the app bundle, config naming/plist changes, and these in the same change?

ihhub commented:
> Fair enough. I was thinking about doing some work to add an app bundle as a build target next, so would it make sense to close this, and then create a new pull request to cover the app bundle, config naming/plist changes/anything else needed to conform to the macOS guidelines, and these in the same change?

Based on detailed explanation of @oleg-derevenetz and if you can help the project with creation of app bundle solution I see no reason to close this pull request for now. We can keep it open and merge it after app bundle related pull request is merged. This will work for cases if we can segregate these 2 sets of changes. I can mark this pull request as draft so we can't merge it by mistake and then later we can unmark it.

What do you think?

Kreeblah commented:
That would work for me, if it's workable for folks here.

ihhub commented on src/engine/tinyconfig.cpp:
```diff
@@ -53,6 +58,107 @@ TinyConfig::TinyConfig( char sep, char com )
 
 bool TinyConfig::Load( const std::string & cfile )
 {
+#if defined( MACOS_APP_BUNDLE )
```
I know that this is WIP pull request but why do we need to have an extra code for reading the configuration file?

ihhub replied:
Even if we need to read / write into a specific file format why can't we get the whole configuration file, compress it and if needed convert into hex. Then later do the opposite operations. My biggest question here do we really need to save a configuration file (essentially we have fheroes2.bin and fheroes2.key files) in such uncommon way?

ihhub replied:
From code support point of view we must avoid adding code specific to some operating systems, especially for configuration files. Let's agree on the way how we should have this file and then proceed with further implementation. We have 2 ways of handling this case:
- use the original fheroes2.cfg file. If we cannot could you please explain why? If we can what's the issue of using it except probably "non-standard" file naming?
- use plist file. Can we save any info there? If yes can we just dump fheroes2.cfg full info there? If not can we use hexed compressed version (using zzlib.h|cpp files)? Converting to HEX is not implemented but this is a very simple code to write.

Kreeblah replied:
From a technical perspective, there is nothing specifically requiring the configuration to be stored as a plist file.  I implemented the plist code to address this requested change to my original submission in this pull request:

> store app preferences in the `~/Library/Preferences` in plist format (instead of simple text config file `fheroes2.cfg`)

Pulling that code back out would be very easy, and the application works just fine that way.  If that is the preferred method here, I can do that.

oleg-derevenetz replied:
Hi all! Regarding this my sentence:

> store app preferences in the ~/Library/Preferences in plist format (instead of simple text config file fheroes2.cfg)

I would like to elaborate a bit. The standard way of storing the app preferences on macOS are plists with special naming scheme. This allows a uniform way of retrieving or updating app preferences, including CLI or scripts - for example, via the built-in macOS `defaults` command:

```bash
defaults write com.apple.dock orientation left
```

But, of course, this means an incompatibility between "app bundle" fheroes2 version and other fheroes2 versions (for example, Linux or Windows) where `fheroes2.cfg` is used. Of course, we still can store preferences in the `fheroes2.cfg`, we will just lose all these additional features in regard to the uniform preferences management on macOS, but in regard to the fheroes2 itself, everything will work. This is up to you, @ihhub :)

ihhub replied:
I am open to have a "proper" MacOS way to save settings but the current approach is not a correct way to do that as we have a lot of duplication code. I will check TinyConfig class and come back with some solution which will allow us to proceed.

oleg-derevenetz replied:
@ihhub This is similar to the "registry vs config file" approaches on Windows BTW :) In fact, we don't use the registry on Windows. so probably it's worth not to use plists on macOS either.

ihhub replied:
I am going to come up with more general solution for settings which will cover cfg, plist or even Windows registry. In this case we can avoid code duplication and maintain the code in a more efficient manner.

ihhub replied:
After further consideration I think it is good for now to stick to one way of saving settings - fheroes2.cfg. We can add a special duplication-free way of saving settings into plist but for now let's stick to fheroes2.cfg file.

Kreeblah replied:
Fair enough.  I just removed the code for the plist files.  I did leave one change, though, and that was the `constexpr` setting the config file name in `src/fheroes2/system/settings.h`.  I can revert that, too, if you'd like, but it seemed like it might make sense to have one place to have that set.

ihhub replied:
It is okay. My apologies for going backward and forward. We are just trying to find the best solution at the moment.

ihhub commented:
Hi @Kreeblah , please let us know when it is ready for review.

Kreeblah commented:
With the question about the plist code resolved, I think this is ready for a review.  The things this pull request now changes are:

- Adds a new build option, `MACOS_APP_BUNDLE`, which is settable as an environment variable for the GNU `make` process, or as a definition for `cmake`
- Changes the search locations for game data (`~/Library/Application Support/fheroes2`), internal data (translations and h2d files, where it searches for them within the app bundle), and configurations (`~/Library/Preferences/fheroes2`) when running an app bundle version
- Added an app bundle `.icns` file
- Added a template for an `Info.plist` file (which describes data about the app bundle), that has key data such as the executable name, version number, etc., substituted by the build process
- Redirected logging output to `syslogd`, which allows it to be viewed in the macOS standard `Console.app` application (note that logging outputs are prefixed with `fheroes2_log` in order to be able to easily find them amongst the other `syslogd` data from the application that's generated by the OS or other libraries used)
- Creates new `zsh` (since `zsh` is the default macOS shell) scripts (`extract_homm2_resources_for_app_bundle.sh` and `download_demo_version_for_app_bundle.sh`) for extracting full game data or downloading demo data to the app bundle-referenced directories
- Switched references to `fheroes2.cfg` to a `constexpr`

Note that this does add a dependency on the CoreFoundation framework for app bundle builds, as I needed to use that in order to determine what the location of the app bundle directory is, for the purposes of finding resources inside of it.  This is included with native macOS build tools, but I don't know whether cross-compilers have anything equivalent.

ihhub commented on CMakeLists.txt:
```diff
@@ -67,27 +71,65 @@ if(NOT WIN32)
-	FILES docs/README.txt LICENSE changelog.txt
-	DESTINATION ${CMAKE_INSTALL_DOCDIR}
-	)
+if(NOT MACOS_APP_BUNDLE)
```
Should we use the opposite logic since we use such in the code above?

Kreeblah replied:
Done.

ihhub commented on script/demo/download_demo_version_for_app_bundle.sh:
```diff
@@ -0,0 +1,73 @@
+#!/bin/zsh
```
The only difference between the old script and this one is the destination path. To remove code duplication I suggest to use an input argument for the original script and if it is not empty use the path provided in the argument. In the newly added script (this one) we can just call the old script with the needed argument. The same applies for the resource extraction scripts. What do you think?

Kreeblah replied:
My concern is that macOS explicitly has deprecated `bash`, and it will probably be removed at some point in the future.  Maybe what I could do is see whether I can create a file to include in both scripts which is valid for both `bash` and `zsh`, and contains the duplicated code.  Would that work?

oleg-derevenetz replied:
@Kreeblah you can just run the main script from the auxiliary script using zsh, something like this:

```sh
#!/bin/zsh

zsh "$(dirname "${(%):-%x}")/name_of_the_main_script.sh"
```

Kreeblah replied:
@oleg-derevenetz - I hadn't thought of that.  The only bashism in the original scripts is the call to `read`, so if I do that in the `zsh` script and pass it as a parameter, it runs fine.

I just changed the scripts to use `zsh` to call the original scripts with additional parameters to specify source/destination locations.  This should remove the code duplication issue.

ihhub commented on src/fheroes2/system/settings.cpp:
```diff
@@ -362,12 +366,17 @@ bool Settings::Save( const std::string & filename ) const
     return true;
 }
 
+std::string Settings::GetConfigFilePayload() const
```
This method is redundant. Please remove it.

ihhub commented on src/fheroes2/system/settings.h:
```diff
@@ -47,6 +47,8 @@ enum MusicSource
 class Settings
 {
 public:
+    static constexpr const char * FHEROES2_CONFIG_FILE_NAME = "fheroes2.cfg";
```
Please name the variable in camelCase variant such as **configFileName** and remove constexpr as it doesn't bring any value here.

oleg-derevenetz replied:
Hi @ihhub I believe in this case `constexpr` is unavoidable because it's the in-class initializer for the static data member of the class. We should either use `constexpr` here or add the out-of-line initialization (in the `.cpp` file).

Kreeblah replied:
I renamed the variable to `configFileName`, however removing the `constexpr` does present a problem, as @oleg-derevenetz noted below.  Removing it results in:

```
error: non-const static data member must be initialized out of line
```

ihhub commented on src/resources/Info.plist.in:
```diff
@@ -0,0 +1,30 @@
+	<key>CFBundlePackageType</key>
+	<string>APPL</string>
+	<key>CFBundleSignature</key>
+	<string>????</string>
```
Could you please explain what it is for?

Kreeblah replied:
CFBundlePackageType indicates what type of thing it is.  APPL indicates an app bundle.

CFBundleSignature is a legacy field.  ???? indicates that there isn't a creator code registered for it, which is something that's only relevant for code capable of running on MacOS 9 or earlier.

ihhub commented on src/resources/Info.plist.in:
```diff
@@ -0,0 +1,30 @@
+	<key>NSPrincipalClass</key>
+	<string>NSApplication</string>
+</dict>
+</plist>
```
This is a question for fheroes2.icns file: can we use any other type of media or only this specific for macOS (?) image type?

Kreeblah replied:
There are a few options for the app icon.  `.icns` is a conglomeration of multiple sizes in one file for display on different pixel density screens, and was the preferred method until recently.

The other possibility for an icon is an [asset catalog](https://developer.apple.com/design/human-interface-guidelines/macos/icons-and-images/app-icon/), which is the current preferred way of doing icons on macOS, but it only supports versions 11 and above.  I'd need to do some research on how to create/manage/use these without an Xcode project if this is the preferred way to go here.

As far as raw image formats, I've never tried them, but I'm not sure they'd work, especially since they wouldn't have the different scaled versions in the single file.  I could give it a try if you'd like, though.

ihhub replied:
My concern is that we already have multiple images for application icons located at `images/platform` and 512x512 page is an upscaled version of another image. We should have a correct image for such case.

Kreeblah replied:
That makes sense.  The one I used to generate the different sizes for the icon file here was `src/resources/fheroes2.png`, which appears to be a 128x128 image, so it got scaled up for the sizes larger than that.

Graphics work is something I don't really have any experience with, though, so I'd need help with this part.  Ideally, there'd be PNGs available at each pixel density (16x16 all the way up to 1024x1024), but even if we just have the largest size, scaling it down is probably fine.

ihhub replied:
First of all, we do not need to provide images for all resolutions at all. If a certain resolution is not present the OS will grab the nearest one and scale it to the required size. Secondly, I've never seen an icon bigger than 256x256 pixels. However, I'm attaching an ICNS file of 512x512 pixels resolution. Please try to use it and see how it looks for the app
[logo_512.zip](https://github.com/ihhub/fheroes2/files/8322074/logo_512.zip)
.

Kreeblah replied:
That looks OK on my machine.  Attached is how it looks in the dock next to a running copy of Preview.app.

I did realize while I was doing this that I needed to add an exclusion for the call to the `setIcon()` method, though, since otherwise that overrides the app bundle icon (and a 32x32 icon doesn't look great).

<img width="126" alt="Screen Shot 2022-03-22 at 01 43 09" src="https://user-images.githubusercontent.com/1812790/159441748-1b97ae88-c7c2-4188-82ae-3ec9008e4940.png">

Kreeblah replied:
Oh, and I replaced the ICNS file in my branch with the one from your ZIP file in the commit I just checked in.

ihhub commented on src/fheroes2/system/settings.cpp:
```diff
@@ -521,6 +529,20 @@ ListDirs Settings::GetRootDirs()
+
+    CFBundleRef mainBundle = CFBundleGetMainBundle();
+    CFURLRef resourcesURL = CFBundleCopyResourcesDirectoryURL( mainBundle );
+    if ( !CFURLGetFileSystemRepresentation( resourcesURL, TRUE, (UInt8 *)resourcePath, PATH_MAX ) ) {
```
Please use C++ static_cast here.

Kreeblah replied:
Done, but I had to use a `reinterpret_cast` rather than a `static_cast`, as `UInt8 *` is essentially an `unsigned char *`.

ihhub commented on src/fheroes2/system/settings.cpp:
```diff
@@ -521,6 +529,20 @@ ListDirs Settings::GetRootDirs()
+    // macOS app bundle Resources directory
+    char resourcePath[PATH_MAX];
+
+    CFBundleRef mainBundle = CFBundleGetMainBundle();
+    CFURLRef resourcesURL = CFBundleCopyResourcesDirectoryURL( mainBundle );
```
We could combine these 2 lines together as we do not use `mainBundle` variable anywhere else:
```cpp
CFURLRef resourcesURL = CFBundleCopyResourcesDirectoryURL( CFBundleGetMainBundle() );
```

Kreeblah replied:
Done.

ihhub commented on src/fheroes2/system/settings.cpp:
```diff
@@ -521,6 +529,20 @@ ListDirs Settings::GetRootDirs()
+    CFBundleRef mainBundle = CFBundleGetMainBundle();
+    CFURLRef resourcesURL = CFBundleCopyResourcesDirectoryURL( mainBundle );
+    if ( !CFURLGetFileSystemRepresentation( resourcesURL, TRUE, (UInt8 *)resourcePath, PATH_MAX ) ) {
+        ERROR_LOG( "Unable to get app bundle path" );
```
Should we **not** to call push_back for this case? For example:
```cpp
if ( CFURLGetFileSystemRepresentation( resourcesURL, TRUE, (UInt8 *)resourcePath, PATH_MAX ) ) {
    dirs.push_back( resourcePath );
}
else {
    ERROR_LOG( "Unable to get app bundle path" );
}
CFRelease( resourcesURL );
```

Kreeblah replied:
Done.

ihhub reviewed (changes requested):
Hi @Kreeblah , I left multiple comments here. Could you please take a look?

ihhub commented on docs/INSTALL.md:
```diff
@@ -107,6 +107,20 @@ If you are using [**Homebrew**](https://brew.sh/), you can install the game by r
+  You will be prompted to enter the location of the original game, and the script will extract all the necessary resource files.
+
+* As an alternative to the previous step, you can manually copy the subdirectories `ANIM`, `DATA`, `MAPS` and `MUSIC` (some of them may
+  be missing depending on the version of the original game) from the original game directory to `~/Library/Application Support/fheroes2`.  Please note that the directories will need to be renamed to lowercase names (`anim`, `data`, `maps`, and `music`).
```
This line contains few places with multiple spaces.

Kreeblah replied:
Done.

oleg-derevenetz commented on src/fheroes2/CMakeLists.txt:
```diff
@@ -10,7 +10,25 @@ get_filename_component(CONFIGURE_FHEROES2_DATA_ABSOLUTE
+	set_target_properties(fheroes2 PROPERTIES MACOSX_BUNDLE_EXECUTABLE_NAME "${MACOSX_BUNDLE_BUNDLE_NAME}")
+	set_target_properties(fheroes2 PROPERTIES MACOSX_BUNDLE_INFO_PLIST "${CMAKE_CURRENT_SOURCE_DIR}/../resources/Info.plist.in")
+	set_target_properties(fheroes2 PROPERTIES MACOSX_BUNDLE_ICON_FILE "fheroes2.icns")
+	set_target_properties(fheroes2 PROPERTIES MACOSX_BUNDLE_GUI_IDENTIFIER "com.horns.and.hoovers.${CMAKE_PROJECT_NAME}")
```
BTW `com.horns.and.hoovers` was a joke :) We have to come up with some kind of identifier. @ihhub any ideas?

ihhub replied:
We could use one of these:
```
com.game.engine.${CMAKE_PROJECT_NAME}
com.game.${CMAKE_PROJECT_NAME}
com.${CMAKE_PROJECT_NAME}
```

oleg-derevenetz replied:
Since the traditional bundle identifier naming scheme is `company_name.app_name`, we could name it as `com.fheroes2.${CMAKE_PROJECT_NAME}`, so we will have `com.fheroes2` as "company name". There will be duplication (resulting name will be `com.fheroes2.fheroes2`), but nothing else comes to mind yet. Suggestions are welcome :)

ihhub replied:
So let's go with `com.fheroes2.${CMAKE_PROJECT_NAME}` for now.

Kreeblah replied:
That's updated now.

oleg-derevenetz commented on src/engine/CMakeLists.txt:
```diff
@@ -1,9 +1,18 @@
+if(MACOS_APP_BUNDLE)
 target_compile_definitions(engine PRIVATE
 	$<$<BOOL:${ENABLE_IMAGE}>:FHEROES2_IMAGE_SUPPORT>
+	MACOS_APP_BUNDLE=1
```
I believe we can just add `$<$<BOOL:${MACOS_APP_BUNDLE}>:MACOS_APP_BUNDLE>` to the `target_compile_definitions` instead of these `if`/`else`/`endif`?

Kreeblah replied:
Done.

oleg-derevenetz commented on src/fheroes2/CMakeLists.txt:
```diff
@@ -10,7 +10,25 @@ get_filename_component(CONFIGURE_FHEROES2_DATA_ABSOLUTE
+	set(fheroes2Icon "${CMAKE_CURRENT_SOURCE_DIR}/../resources/fheroes2.icns")
+	set_source_files_properties(${fheroes2Icon} PROPERTIES MACOSX_PACKAGE_LOCATION "Resources")
+	add_executable(fheroes2 MACOSX_BUNDLE ${fheroes2Icon} ${FHEROES2_SOURCES} ${TRANSLATION_DATA})
+	target_compile_definitions(fheroes2 PRIVATE MACOS_APP_BUNDLE=1)
```
I believe we can just add the `$<$<BOOL:${MACOS_APP_BUNDLE}>:MACOS_APP_BUNDLE>` to the existing `target_compile_definitions` below, isn't it?

Kreeblah replied:
Done.

oleg-derevenetz reviewed (changes requested):
Hi @Kreeblah I left a few suggestions here, please take a look when it's convenient for you :)

Kreeblah commented on docs/INSTALL.md:
````diff
@@ -107,6 +107,20 @@ If you are using [**Homebrew**](https://brew.sh/), you can install the game by r
 brew install fheroes2
 ```
 
+<a name="mac-app-bundle"></a>
````
Done.

oleg-derevenetz commented on docs/INSTALL.md:
```diff
@@ -19,6 +19,7 @@ Precompiled binaries of the release version are currently available for the foll
   * [**MacPorts**](#macports)
   * [**Homebrew**](#homebrew-mac)
   * [**macOS ZIP archive**](#macos-zip-archive)
+  * [**macOS native app**](#macos-native-app)
```
@Kreeblah since "macOS native app" in the text below goes before "macOS ZIP archive", could we please swap these lines here? :)

Kreeblah replied:
Done.

oleg-derevenetz commented on script/homm2/extract_homm2_resources.sh:
```diff
@@ -45,7 +47,7 @@ echo_green "Destination directory: $DEST_PATH"
 echo_stage "[2/3] determining HoMM2 directory"
 
-if [[ "$#" == "1" ]]; then
+if [[ "$#" > "0" ]]; then
```
Since this is an arithmetic comparison, there should be `-gt` instead of `>`. `>` is a string comparison, and, for example, the following script:

```sh
[[ "2" > "10" ]] && echo True
```

will print "True", which is obviously wrong, while the following script:

```sh
[[ "2" -gt "10" ]] && echo True
```

will behave correctly (print nothing).

Kreeblah replied:
That's something I often forget about with shell scripts (the `-gt`/`-lt`/etc. arithmetic comparisons).  Thank you for catching this.  It should be fixed now.

oleg-derevenetz reviewed (changes requested):
Hi @Kreeblah I just left a few more notes, please take a look when it's convenient for you :)

Kreeblah commented on script/demo/download_demo_version.sh:
```diff
@@ -26,7 +26,9 @@ echo_stage "[1/4] determining destination directory"
 DEST_PATH=""
 
-if [[ -f fheroes2 && -x fheroes2 ]]; then
+if [[ ! -z "$1" ]]; then
```
Makes sense to me.

Kreeblah commented on script/homm2/extract_homm2_resources.sh:
```diff
@@ -26,7 +26,9 @@ echo_stage "[1/3] determining destination directory"
 DEST_PATH=""
 
-if [[ -f fheroes2 && -x fheroes2 ]]; then
+if [[ ! -z "$2" ]]; then
```
Same for this.  These should be changed now.

oleg-derevenetz reviewed (approved):
LGTM in general. However, I believe, we still should to come up with app bundle id and deal with app icon.

ihhub commented on src/engine/system.cpp:
```diff
@@ -160,6 +165,12 @@ std::string System::GetDataDirectory( const std::string & prog )
 
+    return std::string();
+#elif defined( MACOS_APP_BUNDLE )
+    if ( getenv( "HOME" ) ) {
```
I missed this place: please cache the output of this function and use the variable for the line below. getenv is an OS call so caching the variable will definitely be faster.

oleg-derevenetz replied:
Hi @ihhub perhaps it is worth doing this in a separate PR for all platforms at once, because it is also not cached for all other platforms? We can just leave a TODO comment for now.

ihhub replied:
Yes. We can do this. I'll do it as I'm making some optimization changes.
